cpp/src/arrow/compute/kernels/vector_rolling.cc

```diff
@@ -0,0 +1,785 @@
+  using OutValue = typename GetOutputType<OutType>::T;
+  KernelContext* ctx;
+  ArgValue prod = 1;
+  explicit ProdCheckedWindow(KernelContext* ctx) : ctx(ctx) {}
+
+  void Append(ArgValue value, Status* st) {
+    prod = MultiplyChecked::Call<ArgValue, ArgValue, ArgValue>(ctx, prod, value, st);
+  }
+  void Remove(ArgValue value, Status* st) {
+    prod = DivideChecked::Call<ArgValue, ArgValue, ArgValue>(ctx, prod, value, st);
+  }
+  OutValue GetValue(Status* st) const { return prod; }
+};
+
+template <typename ArgType>
+struct MinWindow {
+  using OutType = ArgType;
+  using ArgValue = typename GetViewType<ArgType>::T;
+  using OutValue = typename GetOutputType<OutType>::T;
+  KernelContext* ctx;
+  std::multiset<ArgValue> values;
+  explicit MinWindow(KernelContext* ctx) : ctx(ctx) {}
+  void Append(ArgValue value, Status* st) { values.insert(value); }
+  void Remove(ArgValue value, Status* st) {
+    auto it = values.find(value);
+    DCHECK_NE(it, values.end());
+    values.erase(it);
+  }
+  OutValue GetValue(Status* st) const { return *values.begin(); }
+};
+
+template <typename ArgType>
+struct MaxWindow {
+  using OutType = ArgType;
+  using ArgValue = typename GetViewType<ArgType>::T;
+  using OutValue = typename GetOutputType<OutType>::T;
+  KernelContext* ctx;
+  std::multiset<ArgValue, std::greater<ArgValue>> values;
+  explicit MaxWindow(KernelContext* ctx) : ctx(ctx) {}
+  void Append(ArgValue value, Status* st) { values.insert(value); }
+  void Remove(ArgValue value, Status* st) {
+    auto it = values.find(value);
+    DCHECK_NE(it, values.end());
+    values.erase(it);
+  }
+  OutValue GetValue(Status* st) const { return *values.begin(); }
+};
+
+template <typename ArgType>
+struct MeanWindow {
+  using OutType = DoubleType;
+  using ArgValue = typename GetViewType<ArgType>::T;
+  using OutValue = typename GetOutputType<OutType>::T;
+  KernelContext* ctx;
+  double sum = 0;
+  int64_t count = 0;
+  explicit MeanWindow(KernelContext* ctx) : ctx(ctx) {}
+  void Append(ArgValue value, Status* st) {
+    sum += value;
+    count++;
+  }
+  void Remove(ArgValue value, Status* st) {
+    sum -= value;
+    count--;
+  }
+  OutValue GetValue(Status* st) const {
+    DCHECK_GE(count, 0);
+    return sum / count;
+  }
+};
+
+template <typename ArgType, typename Window>
+struct WindowAccumulator {
+  using OutType = typename Window::OutType;
+  using OutValue = typename GetOutputType<OutType>::T;
+  using ArgValue = typename GetViewType<ArgType>::T;
+
+  KernelContext* ctx;
+  const RollingOptions& options;
+  Window window;
+  NumericBuilder<OutType> out_builder;
+  int64_t input_length;
+  WindowAccumulator(KernelContext* ctx, const RollingOptions& options,
+                    int64_t input_length)
+      : ctx(ctx),
+        options(options),
+        window(ctx),
+        out_builder(ctx->memory_pool()),
+        input_length(input_length) {}
+
+  Status AccumulateIgnoreNulls(const ArraySpan& input) {
```

Comment:
Do you think it's possible to pass `IgnoreNulls` as a template parameter and de-duplicate the implementation by adding a few `if constexpr (IgnoreNulls)` in the code?

Reply:
I tried this but there will be so many `if constexpr`s that it hurts readability. The main problem is one needs to keep `window_valid_count` and the other needs `window_null_count`.

Reply:
Makes sense. I've gave up on some `if constexpr`-based implementations in the past because of this as well.